File: examples/javascript/hmc5883l.js
/*
* Author: Zion Orent <zorent@ics.com>
* Copyright (c) 2014 Intel Corporation.
*
* This program and the accompanying materials are made available under the
* terms of the The MIT License which is available at
* https://opensource.org/licenses/MIT.
*
* SPDX-License-Identifier: MIT
*/

// Load Compass module
var hmc5883l = require('jsupm_hmc5883l');
// load this on i2c
var myCompass = new hmc5883l.Hmc5883l(0);
// Set your declination from true north in radians
myCompass.set_declination(0.2749);

var coordsObj;
// Print out the coordinates, heading, and direction every second
setInterval(function()
{
	myCompass.update(); // Update the coordinates
	coordsObj = myCompass.coordinates();
	var coords = (coordsObj.getitem(0) + " ");
	coords += (coordsObj.getitem(1) + " ");
	coords += coordsObj.getitem(2);
	var heading = round2Digits(myCompass.heading());
	var direction = round2Digits(myCompass.direction());
	var compassResults = "coords: " + coords + 
		"    heading: " + heading + " direction " + direction;
	console.log(compassResults);
}, 1000);

function round2Digits(num)
{
	return Math.round((num  + 0.00001) * 100) / 100;
}

// Print message when exiting
process.on('SIGINT', function()
{
	console.log("Exiting...");
	process.exit(0);
});

File: examples/c/hmc5883l.c
/*
 * Author: Jon Trulson <jtrulson@ics.com>
 * Copyright (c) 2017 Intel Corporation.
 *
 * Ported based on original C++ code by:
 * Author: Brendan Le Foll <brendan.le.foll@intel.com>
 * Contributions: Mihai Tudor Panu <mihai.tudor.panu@intel.com>
 * Copyright (c) 2014 Intel Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the The MIT License which is available at
 * https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: MIT
 */

#include <unistd.h>
#include <stdio.h>
#include <signal.h>

#include <upm_utilities.h>

#include "hmc5883l.h"

bool shouldRun = true;

void sig_handler(int signo)
{
  if (signo == SIGINT)
    shouldRun = false;
}

int main(int argc, char **argv)
{
    signal(SIGINT, sig_handler);

//! [Interesting]
    // Instantiate an HMC8883L on I2C bus 0
    hmc5883l_context sensor = hmc5883l_init(0);

    if (!sensor)
    {
        printf("%s: hmc5883l_init() failed\n", __FUNCTION__);
        return 1;
    }

    hmc5883l_set_declination(sensor, 0.2749); // Set your declination
                                              // from true north in
                                              // radians

    // Print out the coordinates, heading, and direction every second
    while (shouldRun)
    {
        hmc5883l_update(sensor); // Update the coordinates
        const int16_t *pos = hmc5883l_coordinates(sensor);

        printf("coor: %5d %5d %5d ", pos[0], pos[1], pos[2]);
        printf("heading: %5.2f direction: %3.2f\n",
               hmc5883l_heading(sensor), hmc5883l_direction(sensor));

        upm_delay(1);
    }

    hmc5883l_close(sensor);

//! [Interesting]

    return 0;
}
